domain/keymanager/state/state.go

```diff
@@ -380,12 +380,79 @@ ON CONFLICT DO NOTHING
 	return err
 }
 
-// GetPublicKeysForUser is responsible for returning all of the public
-// keys for the user uuid on a model. If the user does not exist no error is
-// returned.
-// The following errors can be expected:
-// - [accesserrors.UserNotFound] if the user does not exist.
-// - [modelerrors.NotFound] if the model does not exist.
+// GetAllUsersPublicKeys returns all of the public keys that are in a model and
+// their respective username. This is useful for building a view during model
+// migration. The following errors can be expected:
+// - [modelerrors.NotFound] - When no model exists for the uuid.
+func (s *State) GetAllUsersPublicKeys(
+	ctx context.Context,
+	modelUUID model.UUID,
+) (map[user.Name][]string, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, err
+	}
+
+	modelUUIDVal := modelUUIDValue{UUID: modelUUID.String()}
+
+	stmt, err := s.Prepare(`
+SELECT (u.name, mak.public_key) AS (&userPublicKey.*)
+FROM v_model_authorized_keys AS mak
+INNER JOIN user AS u ON u.uuid = mak.user_uuid
+WHERE mak.model_uuid = $modelUUIDValue.model_uuid
+`, modelUUIDVal, userPublicKey{})
+
+	if err != nil {
+		return nil, errors.Errorf(
+			"preparing select statement for getting all users keys from model %q: %w",
+			modelUUID, err,
+		)
+	}
+	usersKeys := []userPublicKey{}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := s.checkModelExists(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Errorf(
+				"checking model %q exists when getting all users public keys: %w",
+				modelUUID, err,
+			)
+		}
+
+		err = tx.Query(ctx, stmt, modelUUIDVal).GetAll(&usersKeys)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf(
+				"getting all users public keys for model %q: %w", modelUUID, err,
+			)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	rval := map[user.Name][]string{}
+	for _, userKey := range usersKeys {
+		userName, err := user.NewName(userKey.UserName)
```

Comment:
I would argue this is a service layer concern, ie marshalling to/from core package representations.
Just as things like model config are marshalled to/from strings in the service layer, such that the state layer gets exactly what got into the db with no transformation, so too would that apply here. It's the same for other things like Life. The service gets the db lookup value from state and converts to core in the service layer. We would also expect no error since we would have validated what goes into the db is correct, so annotating the error so verbosely seems unnecessary.

Reply:
As discussed f2f, there's a bit to question here, but what's done is consistent with existing code. Therefore any concerns should not block this PR but we should have a discussion about it.